src/lib/telemetry.rs

```diff
@@ -136,12 +136,13 @@ pub enum LogKey {
 
 impl LogKey {
     pub fn add_suffix(&self, suffix: &str) -> LogKey {
-        let s = format!("{}-{}", &self.to_string(), suffix);
-        let s_str = s.as_str();
+        let mut s = self.to_string();
+        s.push_str("-");
+        s.push_str(suffix);
 
-        match LogKey::from_str(s_str) {
+        match LogKey::from_str(&s) {
```

Comment:
The original called `s.as_str()` instead of just using a reference. This is very interesting. I learned that there are a few ways that you can create a string slice "view" into a heap-allocated String. For example, this is all valid Rust:

```rust
// Superfluous type annotations for learning purposes
let s: String = "hello".to_string();
let string_borrowed: &String = &s;
let string_slice: &str = &*s;
let string_slice_another_way: &str = s.as_str();
```

So, borrowing a `String` produces `&String`. However, you can also dereference a `String` using `*`, which produces `str`. This isn't useful because a `str` isn't sized, which means it can't be used without creating a reference to it. So you end up with `&*s`, creating `&str`. This is the functional equivalent of calling `s.as_str()` and is memory efficient because it only creates a "view" into the already existing `String`. This was documented in the first version of the Rust book [here](https://web.mit.edu/rust-lang_v1.25/arch/amd64_ubuntu1404/share/doc/rust/html/book/first-edition/strings.html) (though it doesn't seem to exist in the most recent edition).

However, here is where things get interesting. If you pass a `&String` to a function that calls for `&str`, Rust will coerce it, i.e. dereference it automatically. So you can do this:

```rust
fn coerce(slice: &str) -> &str {
    println!("{}", slice);
    slice
}

fn main() {
    let s: String = "hello".to_string();
    let string_slice: &str = coerce(&s);
}
```

This is a long-winded way of saying that we probably don't need to be calling `as_str()` everywhere.

Reply:
Cool! Thanks for the detailed write-up.